keps/20200901-standard-crd.md

```diff
@@ -0,0 +1,240 @@
+This KEP proposes the CRD Spec and documents the use-cases, not the choice of technology or migration path towards implementing the CRD.
+
+## Terminology
+
+* Kubernetes External Secrets `KES`: A Application that runs a control loop which syncs secrets
+* KES `instance`: A single entity that runs a control loop.
+* ExternalSecret `ES`: A CustomResource that declares which secrets should be synced
+* Store: Is a **source** for secrets. The Store is external to KES. It can be a hosted service like Alibabacloud SecretsManager, AWS SystemsManager, Azure KeyVault...
+* Frontend: A **sink** for the synced secrets. Usually a `ConfigMap` or `Secret`
+* Secret: credentials that act as a key to sensitive information
+
+## Use-Cases
+* one global KES instance that manages ES in **all namespaces**, which gives access to **all stores**, with ACL
+* multiple global KES instances, each manages access to a single or multiple stores (e.g.: shard by stage or team...)
+* one KES per namespace (a user manages his/her own KES instance)
+
+### User definitions
+* `operator :=` i manage one or multiple `KES` instances
+* `user :=` i only create `ES`, KES is managed by someone else
+
+### User Stories
+From that we can derive the following requirements or user-stories:
+1. AS a KES operator i want to run multiple KES instances per cluster (e.g. one KES instance per DEV/PROD)
+2. AS a KES operator or user i want to integrate **multiple stores** from a **single KES instance** (e.g. dev namespace has access only to dev secrets)
+3. AS a KES user i want to control the sink for the secrets (aka frontend: store secret as `kind=ConfigMap` or `kind=Secret`)
+4. AS a KES user i want to fetch **from multiple** stores and store the secrets **in a single** Frontend
+5. AS a KES operator i want to limit the access to certain stores or subresources (e.g. having one central KES instance that handles all ES - similar to `iam.amazonaws.com/permitted` annotation per namespace)
+
+### Stores
+
+These stores are relevant:
+* AWS Secure Systems Manager Parameter Store
+* AWS Secrets Manager
+* Hashicorp Vault
+* Azure Key Vault
+* Alibaba Cloud KMS Secret Manager
+* Google Cloud Platform Secret Manager
+* Kubernetes (see #422)
+* noop (see #476)
+
+### Frontends
+
+* Kind=Secret
+* Kind=ConfigMap
+* *potentially* we could sync store to store
+
+## Proposal
+
+### API
+
+### External Secret
+
+The `ExternalSecret` CustomResourceDefinition is **namespaced**. It defines the following:
+1. source for the secret (store)
+2. sink for the secret (fronted)
+3. and a mapping to translate the keys
+
+```yaml
+apiVersion: kes.io/v1alpha1
+kind: ExternalSecret
+metadata: {...}
+
+spec:
+
+  # optional.
+  # used to select the correct KES instance (think: ingress.ingressClassName)
+  # There is no need for a indirection (e.g. having a extra resource like kind=IngressClass)
+  # the KES controller is instantiated with a specific class name
+  # and filters ES based on this property
+  className: "dev"
```

Comment:
`className` doesn't fit on the ExternalSecret resource because a controller wouldn't be scoped to a set of ExternalSecrets, but rather scoped to a set of SecretStores.

This would fit better as either `className` or `controller` field on the SecretStore type, allowing the controller to only operate on SecretStores (and ExternalSecrets which reference those stores) that the controller is scoped to. This could make using multiple controllers with specific pod access to their stores more simple to setup.

As an example, the AWS ExternalSecret controller with a specific IAM permissions given to the pod would only fetch values for the AWS Secret Manager SecretStore as both the controller and the SecretStore are scoped to the controller class "aws-secret". The Vault ExternalSecret with Kubernetes service account mounted for permissions would only fetch values for Vault SecretStore as both are scoped to the controller class "vault-secret"

This type of SecretStore scoping really only work if ExternalSecrets were limited to referencing a single store, as per my other comment. Otherwise no single controller would have the privileges necessary to create the resulting secret resource.

Reply:
ES is now allowed to reference only a single store. I added a `controller` field to the `SecretStore` type.